File: python/vayu_sdk/apis/api_catalog_products.py
# vayu_sdk/apis/api_catalog_products.py
from openapi.api.catalog_products_api import CatalogProductsApi
from vayu_sdk.clients.vayu_client import VayuClient
from openapi.models.create_catalog_product_request import CreateCatalogProductRequest
from openapi.models.create_catalog_product_response import CreateCatalogProductResponse
from openapi.models.delete_catalog_product_response import DeleteCatalogProductResponse
from openapi.models.get_catalog_product_response import GetCatalogProductResponse
from openapi.models.list_catalog_products_response import ListCatalogProductsResponse
from openapi.models.update_catalog_product_request import UpdateCatalogProductRequest
from openapi.models.update_catalog_product_response import UpdateCatalogProductResponse


class CatalogProductsAPI:
    __client: CatalogProductsApi = None

    def __init__(self, vayu_client: VayuClient):
        self.__client = CatalogProductsApi(vayu_client.client)

    def create(self, request: CreateCatalogProductRequest):
        return self.__client.create_catalog_product(create_catalog_product_request=request)

    def delete(self, id: str):
        return self.__client.delete_catalog_product(catalog_product_id=id)

    def get(self, id: str):
        return self.__client.get_catalog_product(catalog_product_id=id)

    def list(self, limit: int = None, cursor: str = None):
        return self.__client.list_catalog_products(limit=limit, cursor=cursor)

    def update(self, id: str, request: UpdateCatalogProductRequest):
        return self.__client.update_catalog_product(
            update_catalog_product_request=request, catalog_product_id=id
        )

File: go/api/api_catalog_products.go
package api

import (
	"github.com/vayucode/vayu-sdks/go/client"
	"github.com/vayucode/vayu-sdks/go/openapi"
)

type CatalogProduct = openapi.CreateCatalogProductResponseCatalogProduct
type ListCatalogProductsResponse = openapi.ListCatalogProductsResponse
type GetCatalogProductResponse = openapi.GetCatalogProductResponse
type CreateCatalogProductRequest = openapi.CreateCatalogProductRequest
type CreateCatalogProductResponse = openapi.CreateCatalogProductResponse
type UpdateCatalogProductRequest = openapi.UpdateCatalogProductRequest
type UpdateCatalogProductResponse = openapi.UpdateCatalogProductResponse
type DeleteCatalogProductResponse = openapi.DeleteCatalogProductResponse

type CatalogProductsAPI struct {
	vayuClient *client.VayuClient
}

func NewCatalogProductsAPI(vayuClient *client.VayuClient) *CatalogProductsAPI {
	return &CatalogProductsAPI{vayuClient: vayuClient}
}

func NewCreateCatalogProductRequest(name string, description *string, externalId *string) *CreateCatalogProductRequest {
	request := &openapi.CreateCatalogProductRequest{Name: name}
	request.Description = description
	request.ExternalId = externalId
	return request
}

func NewUpdateCatalogProductRequest(name *string, description *string, externalId *string) *UpdateCatalogProductRequest {
	return &openapi.UpdateCatalogProductRequest{Name: name, Description: description, ExternalId: externalId}
}

func (api *CatalogProductsAPI) ListCatalogProducts(limit *float32, cursor *string) (*ListCatalogProductsResponse, error) {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CatalogProductsAPI.ListCatalogProducts(ctx)

	if limit != nil {
		request = request.Limit(*limit)
	}

	if cursor != nil {
		request = request.Cursor(*cursor)
	}

	response, _, err := request.Execute()

	if err != nil {
		return nil, client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return response, nil
}

func (api *CatalogProductsAPI) GetCatalogProduct(catalogProductId string) (*GetCatalogProductResponse, error) {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CatalogProductsAPI.GetCatalogProduct(ctx, catalogProductId)
	response, _, err := request.Execute()

	if err != nil {
		return nil, client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return response, nil
}

func (api *CatalogProductsAPI) CreateCatalogProduct(payload CreateCatalogProductRequest) (*CreateCatalogProductResponse, error) {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CatalogProductsAPI.CreateCatalogProduct(ctx)
	request = request.CreateCatalogProductRequest(payload)
	response, _, err := request.Execute()

	if err != nil {
		return nil, client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return response, nil
}

func (api *CatalogProductsAPI) UpdateCatalogProduct(catalogProductId string, payload UpdateCatalogProductRequest) (*UpdateCatalogProductResponse, error) {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CatalogProductsAPI.UpdateCatalogProduct(ctx, catalogProductId)
	request = request.UpdateCatalogProductRequest(payload)
	response, _, err := request.Execute()

	if err != nil {
		return nil, client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return response, nil
}

func (api *CatalogProductsAPI) DeleteCatalogProduct(catalogProductId string) (*DeleteCatalogProductResponse, error) {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CatalogProductsAPI.DeleteCatalogProduct(ctx, catalogProductId)
	response, _, err := request.Execute()

	if err != nil {
		return nil, client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return response, nil
}
